### PR TITLE
Custom timeout for requests

### PR DESCRIPTION
**Pull request by ckeshava**
## High Level Overview of Change

Certain requests could take more time to return a response, than the default `10 seconds` duration. This change allows users to specify a custom timeout parameter to all requests.
Additionally, this PR sets up uniform interfaces for all `Client._request_impl` overridden functions.

### Context of Change

In the present code, all requests are given 10 seconds duration to return a response. Any longer latency results in a timeout. This is inconvenient for certain commands that are known to take up more time, for instance, `admin`-only commands like the `ledger` command.

### Type of Change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### Did you update CHANGELOG.md?

- [x] Yes
- [ ] No, this change does not impact library users

## Test Plan

This is an existing integration test that invokes the `ledger` command at `tests/integration/reqs/test_ledger.py`. Which other command might take up more time?

### CODE REVIEW

File: xrpl/asyncio/clients/async_websocket_client.py [hunk 1]
```diff
@@ -262,7 +262,9 @@ async def send(self: AsyncWebsocketClient, request: Request) -> None:
             raise XRPLWebsocketException("Websocket is not open")
         await self._do_send(request)
 
-    async def _request_impl(self: WebsocketBase, request: Request) -> Response:
+    async def _request_impl(
+        self: WebsocketBase, request: Request, timeout: Optional[float] = None
```
Comment (mvadari) [inline xrpl/asyncio/clients/async_websocket_client.py:266]:
```suggestion
        self: WebsocketBase, request: Request, timeout: Optional[float] = None
```
This should be a keyword-only argument, to avoid accidental usage.
Ditto everywhere else.

**Review by mvadari — COMMENTED** [review]
It doesn't look like this PR actually modifies the external-facing methods.

File: xrpl/asyncio/clients/client.py [hunk 2]
```diff
@@ -4,9 +4,15 @@
 from abc import ABC, abstractmethod
 from typing import Optional
 
+from typing_extensions import Final
+
 from xrpl.models.requests.request import Request
 from xrpl.models.response import Response
 
+# The default request timeout duration. If you need more time, please pass the required
```
Comment (justinr1234) [inline xrpl/asyncio/clients/client.py:12]: I would reword this to not be addressing someone.

`The default request timeout duration. Set in Client._request_impl to allow more time for longer running commands.`

File: xrpl/asyncio/clients/websocket_base.py [hunk 3]
```diff
@@ -199,7 +199,9 @@ async def _do_pop_message(self: WebsocketBase) -> Dict[str, Any]:
         cast(_MESSAGES_TYPE, self._messages).task_done()
         return msg
 
-    async def _do_request_impl(self: WebsocketBase, request: Request) -> Response:
+    async def _do_request_impl(
+        self: WebsocketBase, request: Request, timeout: Optional[float]
```
Comment (justinr1234) [inline xrpl/asyncio/clients/websocket_base.py:203]: Why is timeout a positional instead of named arg here?

**Review by justinr1234 — APPROVED** [review]

File: xrpl/asyncio/clients/client.py [hunk 4]
```diff
@@ -27,14 +33,22 @@ def __init__(self: Client, url: str) -> None:
...
         async-first. Implement this in a given Client.
 
         Arguments:
             request: An object representing information about a rippled request.
+            timeout: The maximum tolerable delay on waiting for a response.
+                Note: Optional is used in the type in order to honor the existing
+                behavior in certain overridden functions. WebsocketBase.do_request_impl
+                waits indefinitely for the completion of a request, whereas
+                JsonRpcBase._request_impl waits for 10 seconds before timing out a
+                request.
```
[the remark this answers is not in the document]
Comment (mvadari) [reply xrpl/asyncio/clients/client.py:51]: Also, this note shouldn't be in the docstring.

File: xrpl/asyncio/clients/websocket_base.py (hunk 3 above)
Comment (ckeshava) [reply xrpl/asyncio/clients/websocket_base.py:203]: I don't understand, how can I make it a named argument?

File: xrpl/asyncio/clients/client.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ckeshava) [reply xrpl/asyncio/clients/client.py]: addressed in [70dfb80](https://github.com/XRPLF/xrpl-py/pull/687/commits/70dfb80af3b82a90664da5e1b8e6ef60fe409912)

**Comment by ledhed2222** [comment]
i just happened to see this. it looks good to me! great idea :)
